lua/ui/notify/enhancementqueue.lua

```diff
@@ -139,24 +137,27 @@ function ModifyBuildablesForACU(originalBuildables, selection)
 
             ---@param hq UserUnit
             for _, hq in hqs do
+                local bp = hq:GetBlueprint()
+                local hashedCategories = bp.CategoriesHash
                 local supportCategory = categories.SUPPORTFACTORY * factionCategory
 
-                if EntityCategoryContains(categories.TECH3, hq) then
+                if hashedCategories["TECH3"] then
                     supportCategory = supportCategory * (categories.TECH3 + categories.TECH2)
-                elseif EntityCategoryContains(categories.TECH2, hq) then
+                elseif hashedCategories["TECH2"] then
                     supportCategory = supportCategory * categories.TECH2
                 end
 
-                if EntityCategoryContains(categories.LAND, hq) then
-                    supportCategory = supportCategory * categories.LAND
-                elseif EntityCategoryContains(categories.AIR, hq) then
-                    supportCategory = supportCategory * categories.AIR
-                elseif EntityCategoryContains(categories.NAVAL, hq) then
-                    supportCategory = supportCategory * categories.NAVAL
+                if hashedCategories["LAND"] then
+                    supportFactories = supportFactories + supportCategory * categories.LAND
+                end
+                if hashedCategories["AIR"] then
+                    supportFactories = supportFactories + supportCategory * categories.AIR
+                end
+                if hashedCategories["NAVAL"] then
+                    supportFactories = supportFactories + supportCategory * categories.NAVAL
                 end
-                supportFactories = supportFactories + supportCategory
             end
-            newBuildableCategories = supportFactories * supportFactories
+            newBuildableCategories = newBuildableCategories * supportFactories
```

Comment:
The function says it returns a modified list of buildables, not just the valid support factories. Idk why the UI works when you return only support factories.
```suggestion
            newBuildableCategories = newBuildableCategories - categories.SUPPORTFACTORY + supportFactories
```

Reply:
Your suggestion is invalid because It makes all faction HQs to be buildable by any other

Reply:
It doesn't because the original selection is faction specific already.
I found a mistake in my reasoning anyway, I missed the fact that `supportFactories` already starts based off of `newBuildableCategories`, so it includes all original buildables.
Also an issue I found with my suggestion is that it makes T3 support facs buildable by T2 upgrading ACU. The intersection you do here combines the limit from T2 ACU with the limit from available HQs.